macrosynergy/pnl/historic_portfolio_volatility.py

```diff
@@ -170,6 +170,119 @@ def get_max_lookback(lb: int, nt: float) -> int:
     return int(np.ceil(lb * (1 + nt))) if lb > 0 else 0
 
 
+def _calculate_multi_frequency_vcv_for_period(
+    pivot_returns: pd.DataFrame,
+    pivot_signals: pd.DataFrame,
+    rebal_date: pd.Timestamp,
+    est_freqs: List[str],
+    est_weights: List[float],
+    weights_func: Callable[[int, int], np.ndarray],
+    lback_periods: List[int],
+    half_life: List[int],
+    nan_tolerance: float,
+    remove_zeros: bool,
+) -> pd.DataFrame:
+
+    window_df = pivot_returns.loc[pivot_returns.index <= rebal_date]
+    dict_vcv: Dict[str, pd.DataFrame] = {}
+
+    for freq, lb, hl in zip(est_freqs, lback_periods, half_life):
+        piv_ret = _downsample_returns(window_df, freq=freq).iloc[
+            -get_max_lookback(lb=lb, nt=nan_tolerance) :
+        ]
+        dict_vcv[freq] = estimate_variance_covariance(
+            piv_ret=piv_ret,
+            lback_periods=lb,
+            remove_zeros=remove_zeros,
+            weights_func=weights_func,
+            half_life=hl,
+        )
+        # if dict_vcv[freq].isna().any().any():
+        #     raise ValueError(
+        #         f"N/A values in variance-covariance matrix at freq={freq} at real_date={rebal_date}!\n"
+        #         f"{dict_vcv[freq].isna().any()}"
+        #     )
+
+    # NOTE: in this case Float+NA = Na
+    vcv_df: pd.DataFrame = sum(
+        [
+            est_weights[ix] * ANNUALIZATION_FACTORS[freq] * dict_vcv[freq]
+            for ix, freq in enumerate(est_freqs)
+        ]
+    )
+
+    return vcv_df
+
+
+def _calc_vol_tuple(
+    vcv_df: pd.DataFrame,
+    signals: pd.DataFrame,
+    date: pd.Timestamp,
+) -> Tuple[pd.Timestamp, float]:
+    s = signals.loc[date, :].copy()
+
+    # reduce s to vcv columns
+    s = s[s.index.isin(vcv_df.columns)]
```

Comment:
This looks like something that should be an error rather than just filter it out? What is the logic from dropping the signals?

Reply:
by this stage in the program flow, if the VCV matrix has NAs - we have verified either the signal itself is NA/0s, or that the VCV calculation is not possible for that window (due to data availability - NA/0s). Since a calculation cannot be done for a signal with missing VCV data, we can simply drop those columns

Reply:
Sounds reasonable, but requires the check to be in place that the signal is NA/0 before we drop it.